src/protocol/data/resp/parse.c

```diff
@@ -81,67 +83,189 @@ _parse_cmd(struct request *req)
     }
 
-    /* get number of elements in the array */
-    if (!token_is_array(buf)) {
-        log_debug("parse req failed: not an array");
-        return PARSE_EINVALID;
-    }
-    status = token_array_nelem(&nelem, buf);
-    if (status != PARSE_OK) {
-        log_verb("getting array size returns status %d", status);
-        buf->rpos = old_rpos;
-        return status;
-    } else {
-        log_verb("array size is %"PRId64, nelem);
+    /* parse attributes if present */
+    if (token_is_attrib(buf)) {
+        status = parse_element(&attrib, buf);
+        if (status != PARSE_OK) {
+            goto error;
+        }
+        for (;attrib.num > 0; attrib.num--) {
+            status = parse_element(&key, buf);
+            if (status != PARSE_OK) {
+                goto error;
+            }
+            status = parse_element(&val, buf);
+            if (status != PARSE_OK) {
+                goto error;
+            }
+            _parse_req_attrib(req, &key, &val);
+        }
     }
 
-    if (nelem < 1 || nelem > req->token->nalloc) {
-        log_debug("parse req: invalid array size, %d not in [1, %"PRIu32"]",
-                nelem, req->token->nalloc);
-        return PARSE_EINVALID;
+    cap--; /* we will have at least 2 slots here */
+    el = array_push(req->token);
+    status = parse_element(el, buf);
+    if (status != PARSE_OK || el->num < 1) {
+        goto error;
     }
 
+    if (el->type != ELEM_ARRAY) {
+        log_debug("parse req failed: not an array");
+        goto error;
+    }
 
-    /* parse elements */
-    while (nelem > 0) {
-        if (buf_rsize(buf) == 0) {
-            buf->rpos = old_rpos;
-            return PARSE_EUNFIN;
-        }
-        el = array_push(req->token);
-        status = parse_element(el, buf);
-        log_verb("parse element returned status %d", status);
-        if (status != PARSE_OK) {
-            request_reset(req);
-            buf->rpos = old_rpos;
-            return status;
-        }
-        nelem--;
+    status = _parse_range(req->token, buf, el->num);
+    if (status != PARSE_OK) {
+        goto error;
     }
 
     status = _parse_cmd(req);
-    log_verb("parse command returned status %d", status);
     if (status != PARSE_OK) {
-        buf->rpos = old_rpos;
-        return status;
+        goto error;
     }
 
     return PARSE_OK;
+
+error:
+    request_reset(req);
+    buf->rpos = old_rpos;
+    return status;
+}
+
+
+static void
+_parse_rsp_attrib(struct response *rsp, struct element *key, struct element *val)
+{
+    attrib_type_e type = ATTRIB_UNKNOWN;
+
+    /* treat this as it cannot fail, skip anything that is not recognized or has
+     * invalid value so the program will proceed to the request main itself.
+     */
+    if (key->type != ELEM_STR) { /* key has to be a simple string */
+        log_warn("attribute key must be simple string, not type %d", key->type);
+
+        return;
+    }
+    while (++type < ATTRIB_SENTINEL &&
```

Comment:
seems like this is copy pasted from req, maybe we can extract this common logic out into the attribute module?

Reply:
Good call, new local function added.